### ISSUE TITLE
network_ssl might SIGPIPE

### ISSUE THREAD

**Issue by cperciva** [Tarsnap/kivaloo#177]
I don't think OpenSSL has any mechanism for preventing SIGPIPE.  In libcperciva/network we use the POSIX-standard MSG_NOSIGNAL but we can't do that in network_ssl since OpenSSL is handling the socket itself.

@gperciva Can you test to confirm that we have a problem here?  I'd suggest "open a TLS socket and write endless garbage into it, then kill the (TLS) listening daemon" -- the only catch being that you'll need to do something about certificates in order to get the TLS handshake to pass.

Assuming we indeed have an issue here, we probably want to set SO_NOSIGPIPE.  Which doesn't work on Linux... but better to say "TLS sockets aren't available on Linux" than to have utilities which use them randomly crash.

**Comment by gperciva** [comment]
Tentatively confirmed:
```
$ ./nc-client-ssl localhost:4433
asdf
asdf
sdf
$ echo $?
141
```

141 - 128 = 13, which matches SIGPIPE.

I need to double-check my programming, though, because I wasn't expecting to get it with so little text.

**Comment by gperciva** [comment]
Confirmed, and now that I know what I'm looking for, I found other discussions about this, particularly about openssl and non-BSD platforms.

The general solution (i.e. for non-BSD) is `signal(SIGPIPE, SIG_IGN)`.  Which is exactly what you did in `libcperciva/network/network_write.c`:

```
        /* If we don't have MSG_NOSIGNAL, catch SIGPIPE. */
#ifdef POSIXFAIL_MSG_NOSIGNAL
        if ((oldsig = signal(SIGPIPE, SIG_IGN)) == SIG_ERR) {
                warnp("signal(SIGPIPE)");
                goto failed;
        }
#endif
```

Casual testing suggests that we can adopt this for `dowrite()` in `network_ssl.c`, although I guess we can use `#ifndef SO_NOSIGPIPE` there, and set SO_NOSIGPIPE in `network_ssl_open`.

**Comment by gperciva** [comment]
BTW, we can only get `SO_NOSIGPIPE` on FreeBSD if we manually define `__BSD_VISIBLE 1`; it's not normally enabled if we have `_POSIX_C_SOURCE`.  (Details in `/usr/include/sys/cdefs.h`)

**Comment by gperciva** [comment]
Was fixed in #179.
